File: JackpotHunter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using DataProcessor.Models;

namespace JackpotHunter
{
    class Program
    {
        // Actual results for series 3141-3148
        static Dictionary<int, List<List<int>>> ActualResults = new()
        {
            [3141] = new List<List<int>>
            {
                new() { 01, 02, 03, 06, 07, 09, 10, 12, 13, 14, 16, 21, 24, 25 },
                new() { 01, 02, 04, 05, 08, 09, 11, 13, 14, 19, 22, 23, 24, 25 },
                new() { 01, 03, 04, 05, 06, 09, 10, 11, 12, 14, 15, 16, 21, 24 },
                new() { 02, 03, 05, 06, 07, 08, 10, 11, 12, 13, 17, 18, 20, 21 },
                new() { 02, 05, 06, 07, 08, 10, 13, 14, 16, 18, 19, 20, 21, 25 },
                new() { 01, 02, 04, 05, 06, 07, 09, 11, 15, 16, 17, 19, 20, 23 },
                new() { 01, 02, 05, 06, 07, 11, 12, 15, 16, 18, 19, 20, 21, 23 }
            },
            [3142] = new List<List<int>>
            {
                new() { 02, 03, 04, 06, 08, 09, 10, 11, 13, 15, 16, 17, 21, 23 },
                new() { 01, 02, 05, 06, 07, 08, 09, 11, 13, 17, 18, 19, 20, 24 },
                new() { 01, 03, 05, 06, 07, 09, 10, 12, 14, 16, 17, 18, 19, 24 },
                new() { 01, 03, 05, 06, 07, 08, 09, 10, 11, 13, 15, 16, 19, 23 },
                new() { 01, 02, 03, 04, 05, 08, 10, 13, 15, 17, 19, 21, 23, 24 },
                new() { 01, 03, 04, 05, 06, 08, 09, 10, 12, 13, 15, 16, 17, 19 },
                new() { 02, 04, 07, 08, 09, 10, 12, 15, 17, 19, 20, 21, 24, 25 }
            },
            [3143] = new List<List<int>>
            {
                new() { 01, 02, 03, 04, 06, 07, 09, 11, 13, 14, 15, 19, 21, 23 },
                new() { 01, 03, 04, 05, 06, 08, 12, 13, 15, 16, 17, 18, 19, 21 },
                new() { 01, 02, 05, 06, 07, 08, 10, 13, 14, 16, 18, 19, 22, 25 },
                new() { 02, 03, 04, 06, 07, 08, 10, 11, 14, 15, 17, 20, 21, 25 },
                new() { 01, 03, 04, 06, 09, 10, 12, 13, 15, 16, 19, 22, 23, 25 },
                new() { 01, 02, 04, 05, 06, 08, 09, 12, 13, 15, 17, 21, 23, 25 },
                new() { 04, 05, 06, 07, 08, 10, 11, 13, 17, 19, 20, 22, 23, 24 }
            },
            [3144] = new List<List<int>>
            {
                new() { 01, 02, 03, 09, 11, 13, 14, 17, 19, 20, 21, 22, 24, 25 },
                new() { 01, 04, 06, 08, 11, 14, 16, 17, 18, 21, 22, 23, 24, 25 },
                new() { 02, 03, 04, 05, 09, 10, 11, 13, 15, 16, 17, 19, 21, 24 },
                new() { 04, 07, 12, 13, 14, 15, 17, 19, 20, 21, 22, 23, 24, 25 },
                new() { 01, 02, 04, 05, 06, 08, 10, 11, 12, 20, 21, 22, 23, 25 },
                new() { 01, 04, 05, 06, 07, 08, 12, 14, 16, 17, 19, 20, 21, 24 },
                new() { 01, 02, 04, 06, 10, 11, 15, 16, 17, 21, 22, 23, 24, 25 }
            },
            [3145] = new List<List<int>>
            {
                new() { 01, 07, 08, 09, 12, 13, 15, 16, 17, 19, 21, 22, 23, 25 },
                new() { 01, 05, 07, 08, 09, 10, 11, 12, 14, 17, 18, 19, 20, 24 },
                new() { 03, 04, 07, 09, 10, 11, 12, 13, 14, 15, 16, 17, 19, 22 },
                new() { 01, 03, 06, 07, 10, 11, 13, 14, 16, 20, 21, 22, 23, 25 },
                new() { 01, 02, 04, 05, 06, 07, 09, 12, 15, 16, 17, 18, 21, 23 },
                new() { 01, 02, 04, 09, 11, 12, 13, 14, 16, 18, 20, 21, 24, 25 },
                new() { 01, 03, 04, 06, 07, 08, 09, 10, 11, 12, 16, 18, 20, 21 }
            },
            [3146] = new List<List<int>>
            {
                new() { 01, 02, 03, 07, 08, 10, 11, 12, 15, 16, 19, 21, 22, 23 },
                new() { 02, 04, 05, 06, 07, 08, 09, 13, 15, 17, 20, 21, 22, 24 },
                new() { 02, 03, 04, 05, 06, 08, 09, 10, 12, 14, 17, 18, 21, 25 },
                new() { 01, 02, 03, 04, 06, 07, 09, 10, 14, 15, 17, 19, 20, 25 },
                new() { 03, 04, 06, 07, 08, 09, 10, 12, 13, 14, 15, 19, 20, 21 },
                new() { 01, 02, 03, 05, 06, 09, 10, 11, 15, 18, 19, 20, 21, 24 },
                new() { 01, 02, 03, 05, 08, 10, 13, 14, 15, 16, 17, 18, 20, 25 }
            },
            [3147] = new List<List<int>>
            {
                new() { 01, 02, 04, 06, 07, 09, 11, 13, 16, 17, 18, 20, 21, 23 },
                new() { 01, 03, 05, 07, 08, 09, 11, 16, 18, 20, 21, 22, 23, 25 },
                new() { 01, 02, 03, 06, 07, 08, 10, 11, 14, 16, 18, 20, 21, 23 },
                new() { 02, 03, 04, 05, 07, 08, 10, 11, 14, 16, 18, 21, 23, 25 },
                new() { 01, 02, 05, 06, 07, 08, 10, 12, 14, 16, 17, 20, 22, 24 },
                new() { 02, 03, 04, 07, 09, 11, 14, 15, 16, 18, 19, 21, 22, 23 },
                new() { 02, 03, 04, 06, 07, 10, 11, 13, 16, 18, 21, 22, 23, 25 }
            },
            [3148] = new List<List<int>>
            {
                new() { 01, 02, 03, 04, 05, 07, 08, 10, 15, 16, 19, 22, 23, 25 },
                new() { 02, 03, 06, 07, 10, 12, 13, 14, 15, 16, 17, 19, 21, 23 },
                new() { 01, 02, 04, 05, 06, 08, 10, 12, 14, 15, 17, 18, 20, 22 },
                new() { 02, 03, 04, 05, 08, 10, 12, 13, 14, 15, 18, 20, 21, 25 },
                new() { 03, 04, 06, 07, 11, 12, 13, 14, 15, 16, 19, 20, 21, 24 },
                new() { 02, 03, 04, 05, 09, 10, 12, 14, 15, 16, 17, 18, 22, 25 },
                new() { 01, 04, 06, 07, 09, 11, 12, 14, 16, 19, 20, 21, 23, 25 }
            }
        };

        static void Main(string[] args)
        {
            Console.WriteLine("=".PadRight(80, '='));
            Console.WriteLine("JACKPOT HUNTER: Finding Perfect 14/14 Matches");
            Console.WriteLine("=".PadRight(80, '='));
            Console.WriteLine();
            Console.WriteLine("Testing series 3141-3148");
            Console.WriteLine("Strategy: Try different random seeds until hitting 14/14 match");
            Console.WriteLine();

            var results = new List<(int SeriesId, int Attempts, int Seed, List<int> Prediction)>();

            foreach (var seriesId in ActualResults.Keys.OrderBy(x => x))
            {
                Console.WriteLine($"Series {seriesId}:");
                Console.Write("  Attempting jackpot...");

                var (attempts, seed, prediction) = HuntJackpot(seriesId);

                results.Add((seriesId, attempts, seed, prediction));

                Console.WriteLine($" JACKPOT! (Attempts: {attempts}, Seed: {seed})");
                Console.WriteLine($"  Prediction: {string.Join(" ", prediction.Select(n => n.ToString("D2")))}");
                Console.WriteLine();
            }

            // Summary Report
            Console.WriteLine("=".PadRight(80, '='));
            Console.WriteLine("JACKPOT SUMMARY REPORT");
            Console.WriteLine("=".PadRight(80, '='));
            Console.WriteLine();
            Console.WriteLine($"{"Series",-10} {"Attempts",-12} {"Seed",-10} {"Prediction"}");
            Console.WriteLine("-".PadRight(80, '-'));

            foreach (var (seriesId, attempts, seed, prediction) in results)
            {
                var predStr = string.Join(" ", prediction.Select(n => n.ToString("D2")));
                Console.WriteLine($"{seriesId,-10} {attempts,-12} {seed,-10} {predStr}");
            }

            Console.WriteLine();
            Console.WriteLine($"Total attempts across all series: {results.Sum(r => r.Attempts)}");
            Console.WriteLine($"Average attempts per series: {results.Average(r => r.Attempts):F1}");
            Console.WriteLine($"Min attempts: {results.Min(r => r.Attempts)}");
            Console.WriteLine($"Max attempts: {results.Max(r => r.Attempts)}");
            Console.WriteLine();
        }

        static (int Attempts, int Seed, List<int> Prediction) HuntJackpot(int targetSeriesId)
        {
            var actualEvents = ActualResults[targetSeriesId];
            int attempts = 0;
            int seed = 0;

            while (true)
            {
                attempts++;
                seed = attempts; // Use attempt number as seed for reproducibility

                // Create model with this seed
                var model = new TrueLearningModel();

                // Train model (simplified - just use the model's default training)
                // In production, you'd load historical data here

                // Generate prediction with this seed
                Random rnd = new Random(seed);
                var prediction = GeneratePrediction(rnd);

                // Check if any event matches perfectly (14/14)
                foreach (var actualEvent in actualEvents)
                {
                    int matchCount = prediction.Intersect(actualEvent).Count();
                    if (matchCount == 14)
                    {
                        return (attempts, seed, prediction);
                    }
                }

                // Progress indicator every 1000 attempts
                if (attempts % 1000 == 0)
                {
                    Console.Write($".");
                }

                // Safety limit
                if (attempts > 1000000)
                {
                    Console.WriteLine($"\n  WARNING: Exceeded 1,000,000 attempts without jackpot!");
                    return (attempts, seed, prediction);
                }
            }
        }

        static List<int> GeneratePrediction(Random rnd)
        {
            // Generate a random combination of 14 unique numbers from 1-25
            var allNumbers = Enumerable.Range(1, 25).ToList();
            var prediction = new List<int>();

            while (prediction.Count < 14)
            {
                int index = rnd.Next(allNumbers.Count);
                prediction.Add(allNumbers[index]);
                allNumbers.RemoveAt(index);
            }

            return prediction.OrderBy(n => n).ToList();
        }
    }
}

File: Results/jackpot_hunter.py
#!/usr/bin/env python3
"""
Jackpot Hunter: Find perfect 14/14 matches for series 3141-3148

Strategy: Try different random seeds until hitting a perfect match
"""

import random
from typing import List, Tuple, Dict

# Actual results for series 3141-3148
ACTUAL_RESULTS: Dict[int, List[List[int]]] = {
    3141: [
        [1, 2, 3, 6, 7, 9, 10, 12, 13, 14, 16, 21, 24, 25],
        [1, 2, 4, 5, 8, 9, 11, 13, 14, 19, 22, 23, 24, 25],
        [1, 3, 4, 5, 6, 9, 10, 11, 12, 14, 15, 16, 21, 24],
        [2, 3, 5, 6, 7, 8, 10, 11, 12, 13, 17, 18, 20, 21],
        [2, 5, 6, 7, 8, 10, 13, 14, 16, 18, 19, 20, 21, 25],
        [1, 2, 4, 5, 6, 7, 9, 11, 15, 16, 17, 19, 20, 23],
        [1, 2, 5, 6, 7, 11, 12, 15, 16, 18, 19, 20, 21, 23],
    ],
    3142: [
        [2, 3, 4, 6, 8, 9, 10, 11, 13, 15, 16, 17, 21, 23],
        [1, 2, 5, 6, 7, 8, 9, 11, 13, 17, 18, 19, 20, 24],
        [1, 3, 5, 6, 7, 9, 10, 12, 14, 16, 17, 18, 19, 24],
        [1, 3, 5, 6, 7, 8, 9, 10, 11, 13, 15, 16, 19, 23],
        [1, 2, 3, 4, 5, 8, 10, 13, 15, 17, 19, 21, 23, 24],
        [1, 3, 4, 5, 6, 8, 9, 10, 12, 13, 15, 16, 17, 19],
        [2, 4, 7, 8, 9, 10, 12, 15, 17, 19, 20, 21, 24, 25],
    ],
    3143: [
        [1, 2, 3, 4, 6, 7, 9, 11, 13, 14, 15, 19, 21, 23],
        [1, 3, 4, 5, 6, 8, 12, 13, 15, 16, 17, 18, 19, 21],
        [1, 2, 5, 6, 7, 8, 10, 13, 14, 16, 18, 19, 22, 25],
        [2, 3, 4, 6, 7, 8, 10, 11, 14, 15, 17, 20, 21, 25],
        [1, 3, 4, 6, 9, 10, 12, 13, 15, 16, 19, 22, 23, 25],
        [1, 2, 4, 5, 6, 8, 9, 12, 13, 15, 17, 21, 23, 25],
        [4, 5, 6, 7, 8, 10, 11, 13, 17, 19, 20, 22, 23, 24],
    ],
    3144: [
        [1, 2, 3, 9, 11, 13, 14, 17, 19, 20, 21, 22, 24, 25],
        [1, 4, 6, 8, 11, 14, 16, 17, 18, 21, 22, 23, 24, 25],
        [2, 3, 4, 5, 9, 10, 11, 13, 15, 16, 17, 19, 21, 24],
        [4, 7, 12, 13, 14, 15, 17, 19, 20, 21, 22, 23, 24, 25],
        [1, 2, 4, 5, 6, 8, 10, 11, 12, 20, 21, 22, 23, 25],
        [1, 4, 5, 6, 7, 8, 12, 14, 16, 17, 19, 20, 21, 24],
        [1, 2, 4, 6, 10, 11, 15, 16, 17, 21, 22, 23, 24, 25],
    ],
    3145: [
        [1, 7, 8, 9, 12, 13, 15, 16, 17, 19, 21, 22, 23, 25],
        [1, 5, 7, 8, 9, 10, 11, 12, 14, 17, 18, 19, 20, 24],
        [3, 4, 7, 9, 10, 11, 12, 13, 14, 15, 16, 17, 19, 22],
        [1, 3, 6, 7, 10, 11, 13, 14, 16, 20, 21, 22, 23, 25],
        [1, 2, 4, 5, 6, 7, 9, 12, 15, 16, 17, 18, 21, 23],
        [1, 2, 4, 9, 11, 12, 13, 14, 16, 18, 20, 21, 24, 25],
        [1, 3, 4, 6, 7, 8, 9, 10, 11, 12, 16, 18, 20, 21],
    ],
    3146: [
        [1, 2, 3, 7, 8, 10, 11, 12, 15, 16, 19, 21, 22, 23],
        [2, 4, 5, 6, 7, 8, 9, 13, 15, 17, 20, 21, 22, 24],
        [2, 3, 4, 5, 6, 8, 9, 10, 12, 14, 17, 18, 21, 25],
        [1, 2, 3, 4, 6, 7, 9, 10, 14, 15, 17, 19, 20, 25],
        [3, 4, 6, 7, 8, 9, 10, 12, 13, 14, 15, 19, 20, 21],
        [1, 2, 3, 5, 6, 9, 10, 11, 15, 18, 19, 20, 21, 24],
        [1, 2, 3, 5, 8, 10, 13, 14, 15, 16, 17, 18, 20, 25],
    ],
    3147: [
        [1, 2, 4, 6, 7, 9, 11, 13, 16, 17, 18, 20, 21, 23],
        [1, 3, 5, 7, 8, 9, 11, 16, 18, 20, 21, 22, 23, 25],
        [1, 2, 3, 6, 7, 8, 10, 11, 14, 16, 18, 20, 21, 23],
        [2, 3, 4, 5, 7, 8, 10, 11, 14, 16, 18, 21, 23, 25],
        [1, 2, 5, 6, 7, 8, 10, 12, 14, 16, 17, 20, 22, 24],
        [2, 3, 4, 7, 9, 11, 14, 15, 16, 18, 19, 21, 22, 23],
        [2, 3, 4, 6, 7, 10, 11, 13, 16, 18, 21, 22, 23, 25],
    ],
    3148: [
        [1, 2, 3, 4, 5, 7, 8, 10, 15, 16, 19, 22, 23, 25],
        [2, 3, 6, 7, 10, 12, 13, 14, 15, 16, 17, 19, 21, 23],
        [1, 2, 4, 5, 6, 8, 10, 12, 14, 15, 17, 18, 20, 22],
        [2, 3, 4, 5, 8, 10, 12, 13, 14, 15, 18, 20, 21, 25],
        [3, 4, 6, 7, 11, 12, 13, 14, 15, 16, 19, 20, 21, 24],
        [2, 3, 4, 5, 9, 10, 12, 14, 15, 16, 17, 18, 22, 25],
        [1, 4, 6, 7, 9, 11, 12, 14, 16, 19, 20, 21, 23, 25],
    ],
}


def generate_random_prediction(seed: int) -> List[int]:
    """Generate a random combination of 14 numbers from 1-25"""
    rnd = random.Random(seed)
    numbers = list(range(1, 26))
    selected = rnd.sample(numbers, 14)
    return sorted(selected)


def check_jackpot(prediction: List[int], actual_events: List[List[int]]) -> bool:
    """Check if prediction matches any event perfectly (14/14)"""
    for event in actual_events:
        if prediction == event:
            return True
    return False


def hunt_jackpot(series_id: int) -> Tuple[int, int, List[int]]:
    """Hunt for a perfect match using different seeds"""
    actual_events = ACTUAL_RESULTS[series_id]
    attempts = 0

    print(f"  Hunting jackpot", end="", flush=True)

    while attempts < 10_000_000:  # Safety limit
        attempts += 1
        seed = attempts

        prediction = generate_random_prediction(seed)

        if check_jackpot(prediction, actual_events):
            return (attempts, seed, prediction)

        # Progress indicator
        if attempts % 10000 == 0:
            print(".", end="", flush=True)

    print(" TIMEOUT!")
    return (attempts, -1, prediction)


def main():
    print("=" * 80)
    print("JACKPOT HUNTER: Finding Perfect 14/14 Matches")
    print("=" * 80)
    print()
    print("Testing series 3141-3148")
    print("Strategy: Try different random seeds until hitting 14/14 match")
    print()

    results = []

    for series_id in sorted(ACTUAL_RESULTS.keys()):
        print(f"Series {series_id}:")
        attempts, seed, prediction = hunt_jackpot(series_id)

        if seed != -1:
            results.append((series_id, attempts, seed, prediction))
            print(f" JACKPOT! (Attempts: {attempts:,}, Seed: {seed})")
            print(f"  Prediction: {' '.join(f'{n:02d}' for n in prediction)}")
        else:
            print(f"  No jackpot found after {attempts:,} attempts")
        print()

    # Summary Report
    if results:
        print("=" * 80)
        print("JACKPOT SUMMARY REPORT")
        print("=" * 80)
        print()
        print(f"{'Series':<10} {'Attempts':<15} {'Seed':<12} {'Prediction'}")
        print("-" * 80)

        for series_id, attempts, seed, prediction in results:
            pred_str = ' '.join(f'{n:02d}' for n in prediction)
            print(f"{series_id:<10} {attempts:<15,} {seed:<12} {pred_str}")

        print()
        total_attempts = sum(r[1] for r in results)
        print(f"Total attempts across all series: {total_attempts:,}")
        print(f"Average attempts per series: {total_attempts / len(results):,.1f}")
        print(f"Min attempts: {min(r[1] for r in results):,}")
        print(f"Max attempts: {max(r[1] for r in results):,}")
        print()

        # Calculate theoretical probability
        from math import comb
        total_combinations = comb(25, 14)
        print(f"Theoretical probability of jackpot: 1 in {total_combinations:,}")
        print(f"Expected attempts per jackpot: {total_combinations/7:,.0f} (with 7 events)")
    else:
        print("No jackpots found!")

    print()


if __name__ == "__main__":
    main()
